### DISCUSSION TITLE
Entanglement via this.$wire does not work

### DISCUSSION THREAD

**Discussion by IP-Developer** [livewire/livewire#3051] [7. Bugs (archived)]
### Description
Entangling Alpine properties using `this.$wire` doesn't seem to work. The Proxy object is created, but the value is never retrieved or set. To add to this, entanglement doesn't work with `@entangle` directive if set in x-init. The only way entanglement works is when setting with `@entangle` directly on property definition, which is pretty limiting, considering that all my Alpine components are bundled with Mix.

### Exact steps to reproduce
[Laravel Playground](https://laravelplayground.com/#/snippets/6d02ebac-f251-41b0-8b0a-3b92479d32fa)

**Context**
- Livewire version: [e.g. 2.4.1]
- Laravel version: [e.g. 8.32.1]
- Alpine version: [e.g. 2.8.2]
- Browser: Any

**Comment by zepfietje** [comment]
I believe this has been an issue for quite some time already.

**Comment by joshhanley** [comment]
@IP-Developer using `@entangle` only works inside `x-data` as Livewire parses it before the Alpine component is initialised. It does that by removing the entangle call, sets up watchers, and then assigns the Livewire property value to the Alpine property.

So trying to use entangle in `x-init` won't work.

But I'm not exactly clear on why you would need that. Any chance you can explain how/why you would use that, and what you mean by your Alpine components are bundled with Mix and why that stops you using entangle in `x-data`?

Thanks!

**Comment by IP-Developer** [comment]
@joshhanley I am aware that `@entangle` works. The problem is specifically with entanglement via `this.$wire.entangle()`. This function is documented, and by my judgement, there shouldn't be any problem with using it, as Alpine components always load after Livewire. Please correct me if I'm wrong (in that case it would be better to wrap `this.$wire.entangle()` in livewire hook?).

The reason I want to stick to `this.$wire.entangle()` is because all my Alpine components are bundled with Laravel Mix (and are pretty big actually), so I don't have access to the `@entangle` directive. An option would be to pass entanglements via `@entangle` directly to Alpine component constructor function, but this just creates more confusion as it's not self-explanatory in the JS file. So the best way to achieve that would be to use `this.$wire.entangle()` in an `init()` component method called by `x-init` (check the playground). I'm trying to avoid having inline scripts inside Livewire components as those are not cached or transpiled, and also makes it harder to third-party packages, as it forces me to assign everything to `window` so it can be accessible.

**Comment by joshhanley** [comment]
@IP-Developer `@entangle` just expands to `$wire.entangle()` so you can use `$wire.entangle()` in `x-data` without any issue but it won't work in `x-init`.

If it is documented as `this.$wire.entangle()` then that is an error, can you show me where that is so we can get it fixed?

If you are curious, this is the logic which parses entangle inside `x-data` which happens before the Alpine component is initialised (hence why won't work in init) https://github.com/livewire/livewire/blob/2f06b28be0aa0d5bf0a4598b4b571d7d1a24e5c1/js/component/SupportAlpine.js#L51

Ok thanks for sharing, that makes sense, but yeah sadly that won't work. 

To dynamically assign entangle would require changing the way `entangle` is initialised.

I would personally just pass through the Alpine component constructor and maybe just document that, that is how it's running if that is confusing for others.

I use something like this

```html
<div x-data="something({ value: $wire.entangle('livewireValue') })">
```

```js
function something(config) {
  return {
    ...config,
    // other stuff
  }
}
```

But agree using the `...config` makes it harder to reason about what properties are available in the component. So sometime I use this instead

```js
function something(config) {
  return {
    value: config.value,
    // other stuff
  }
}
```

Hope this helps!

**Comment by IP-Developer** [comment]
@joshhanley to answer your last sentence, my Alpine components' logic are extracted to functions [(check x-data documentation)](https://github.com/alpinejs/alpine#x-data). All those functions are defined in my `app.js` file, which is then bundled/transpiled/polyfilled by Laravel Mix. A simple example of an extracted Alpine component in `app.js`:
```js
window.myComponent = () => 
{
    return { 
        name: 'test',
        init(name) 
        {
            this.name = name;
        }
    }
}
```
which is then used like so:
```html
<div x-data="myComponent()" x-init="init('something else')"></div>
```

So the exact thing I was trying to do is both having `name` property defined in Alpine component with a default value, and then entangle it to Livewire's property in `init()`.

**Comment by IP-Developer** [comment]
@joshhanley 

> If it is documented as `this.$wire.entangle()` then that is an error, can you show me where that is so we can get it fixed?

It is indeed, check the [Reference](https://laravel-livewire.com/docs/2.x/reference) page in Livewire's documentation.

I think getting it to work inside `x-init` would be a huge improvement, considering that Alpine components definition via Alpine.component() is coming in v3 (also magic __init() method), I think most developers will switch to defining complex components inside `app.js` and even loading them dynamically via webpack's `import()`.

Maybe passing `$wire` inside component function could be an option? So something like this would work:
```js
function myComponent($wire) 
{
    return {
        name: $wire.entangle('name')
    }
}
```

**Comment by joshhanley** [comment]
@IP-Developer ah yeah, thanks for the link! Yeah the option on that page `$wire.entangle('foo').defer` should actually be wrapped in `x-data` so it is clearer that is how it should be used, like the `x-data="{ foo: $wire.entangle('foo') }"` above it.

Yeah I think dynamic entangle assignment could be helpful.

So, I thought just using `name: this.$wire.entangle('name')` would work but it comes up with a `$wire` undefined error. But then I remembered using `name: @entangle('name')` works.

But looking at the expanded code for `@entangle`, it actually expands to `window.Livewire.find('<livewire-component-id>').entangle('name')`

So maybe instead of passing $wire through the constructor, you instead pass the Livewire component through using `@this` (which expands to `window.Livewire.find('...')`) and then make use of that Livewire object yourself like

```html
<div x-data="something(@this)">
```

```js
function myComponent(livewireComponent) 
{
    return {
        name: livewireComponent.entangle('name')
    }
}
```

Just tested it and it works. Hopefully that would make it clearer what is going on.

**Reply by momojahlow** [reply]
Thanks a lot... it works for me

**Comment by IP-Developer** [comment]
@joshhanley yes, this is exactly what I came up to right now :) But `this.$wire.entangle()` still seems to be the best option for such cases (considering upcoming changes in Alpine, it would be perfect). P.S. Thanks for your help!

@calebporzio any thoughts on this?

**Comment by joshhanley** [comment]
@IP-Developer no worries 🙂

**Comment by rizkhal** [comment]
did alpinejs 3 remove entangle? i use alpinejs v3 and `@entangle` does not work but works with alpinejs v2
below is error log when trying to use `@entangle` on v3

![Screenshot from 2021-08-20 22-59-37](https://user-images.githubusercontent.com/24653114/130244521-99ef814d-a470-4d61-a237-b01a981617f7.png)

**Reply by joshhanley** [reply]
@rizkhal no, works fine for me.

Make sure you are on the latest version of Livewire though, as older versions aren't compatible with V3, as some changes had to be made. Latest is 2.5.5.

Can you please upgrade and let us know how it goes?

**Reply by rizkhal** [reply]
Hy @joshhanley, after update alpinejs from v2 to v3, i have the error on the picture below.

i remove alpine installation from npm and put the alpinejs v3 cdn on my layout (inside head tag)
```
    <script defer src="https://unpkg.com/alpinejs@3.2.3/dist/cdn.min.js"></script>
```

i update livewire from 2.x to 2.5.5 and also clear cache.

![Screenshot from 2021-08-22 21-01-17](https://user-images.githubusercontent.com/24653114/130354269-3a5c7b38-eb6c-45a3-b4d0-43e29845d47a.png)

Thanks

**Reply by rizkhal** [reply]
Oh sorry my mistake, this error because on my custom event
on v2 i have listener like
```js
<div @alert.window="const something = doSomething () => {}">
 something else
</div>
```
that's solved after i update like

```js
<div @alert.window="() => {
  const something = doSomething () => {}
}">
 something else
</div>
```

`@entangle` works and everything is works fine, Thanks @joshhanley

**Reply by joshhanley** [reply]
@rizkhal no worries 🙂 glad you worked it out

**Comment by phuhung273** [comment]
Can't make entangle works. I have tried all of your solution @joshhanley :
```
<div x-data="something (@this)" />

<script>
  document.addEventListener('alpine:init', () => {
      Alpine.data('something', (livewireComponent) => ({
          name: livewireComponent.entangle('name'),
      }))
  })
</script>
```
```
<div x-data="something({ name: $wire.entangle('name') })" />

<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('something', (config) => ({
            ...config
        }))
    })
</script>
```
All lead to $wire undefined.

I use Alpine 3. @joshhanley @IP-Developer @rizkhal please help me ?

**Reply by joshhanley** [reply]
@phuhung273 can you share your layout file and if you are adding Alpine in through NPM, then can you also share your `app.js` file where you include Alpine?

Also can you share specifically what patch versions of Livewire and Alpine you are using?

Thanks!

**Reply by phuhung273** [reply]
form.blade.php
```
<div>
  <form wire:submit.prevent="submit" enctype="multipart/form-data">
          <x-image-upload wire:model.defer="image" />
  </form>
</div>
```
Form.php
```
class Form extends Component
{
    use WithFileUploads;

    public $image;
    public $isPreview;

    public function submit()
    {
        // ...Do stuff

        $this->resetForm();
    }

    private function resetForm() {
        $this->image = null;
        $this->isPreview = false;
    }
}
```
image-upload.blade.php
```
<div x-data="imageUploadData(@this)" class="col-span-6 ml-2 sm:col-span-4 md:mr-3">
    <!-- Photo File Input -->
    <input {{ $attributes }} type="file" x-ref="photo" class="hidden" x-on:change="changeImage">

    <div class="text-center">
        <!-- Current Photo -->
        <div class="mt-2" x-show="!isPreview">
            <div class="w-10/12 m-auto">
                <img class="object-cover" src="https://images.unsplash.com/photo-1531316282956-d38457be0993?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=700&q=80" />
            </div>
        </div>
        <!-- New Photo Preview -->
        <div class="mt-2" x-show="isPreview">
            <span class="block w-10/12 h-40 m-auto shadow"
                x-bind:style="'background-size: cover; background-repeat: no-repeat; background-position: center center; background-image: url(\'' + photoPreview + '\');'"
                style="background-size: cover; background-repeat: no-repeat; background-position: center center;">
            </span>
        </div>
        <button type="button"
            class="inline-flex items-center px-4 py-2 bg-white border border-gray-300 rounded-md font-semibold text-xs text-gray-700 uppercase tracking-widest shadow-sm hover:text-gray-500 focus:outline-none focus:border-blue-400 focus:shadow-outline-blue active:text-gray-800 active:bg-gray-50 transition ease-in-out duration-150 mt-2"
            x-on:click.prevent="$refs.photo.click()">
            Select New Photo
        </button>
    </div>
</div>

@push('scripts')
<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('imageUploadData', (livewireComponent) => ({
            isPreview: livewireComponent.entangle('isPreview'),
            photoPreview: null,
            changeImage(e){
                const reader = new FileReader();
                reader.onload = (e) => {
                    this.photoPreview = e.target.result;
                    this.isPreview = true;
                };
                reader.readAsDataURL(e.target.files[0]);
            }
        }))
    })
</script>
@endpush
```
Version:
- Laravel: 8.0
- Livewire: 2.0
- Alpinejs: 3.2.4

Sorry for the last snippet, I just want to show it full in case you want something visible.

Thanks for catching up @joshhanley

**Reply by joshhanley** [reply]
@phuhung273 can you share your layout file and app.js (if you are using) as I request please 🙂

Also can you share your exact patch versions please? Run `composer show -D` in your terminal to get the installed versions.
